From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: killerprojecte <admin@fastmcmirror.org>
Date: Mon, 12 Jun 2023 19:05:33 +0800
Subject: [PATCH] Recode DirtyFolia


diff --git a/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java b/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
index 02ab82f6d0eb2a5afbb6b0ac4d738097c4986ad1..cff0773df89752cfedffec04e423193d85f897e1 100644
--- a/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
+++ b/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
@@ -269,6 +269,9 @@ public final class PluginDescriptionFile implements io.papermc.paper.plugin.conf
      */
     @Override
     public boolean isFoliaSupported() {
+        if (System.getProperty("folia.force-load-all-plugins", "false").equalsIgnoreCase("true")) {
+            return true;
+        }
         return this.foliaSupported != null && this.foliaSupported.equalsIgnoreCase("true");
     }
     // Folia end - block plugins not marked as supported
