From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: killerprojecte <admin@fastmcmirror.org>
Date: Tue, 13 Jun 2023 22:02:16 +0800
Subject: [PATCH] Update implementation version


diff --git a/build.gradle.kts b/build.gradle.kts
index 9ee90bc753e65f9dd532f28ec75cc16bf16486c8..00a6c57714e23ac00a66fdba5cca2b49c6ff4775 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -76,7 +76,7 @@ tasks.jar {
         attributes(
             "Main-Class" to "org.bukkit.craftbukkit.Main",
             "Implementation-Title" to "CraftBukkit",
-            "Implementation-Version" to "git-Folia-$implementationVersion", // Folia
+            "Implementation-Version" to "git-DirtyFolia-$implementationVersion", // Folia
             "Implementation-Vendor" to date, // Paper
             "Specification-Title" to "Bukkit",
             "Specification-Version" to project.version,
