From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: killerprojecte <admin@fastmcmirror.org>
Date: Mon, 19 Jun 2023 14:57:27 +0800
Subject: [PATCH] Cache world data when it not null


diff --git a/src/main/java/net/minecraft/world/level/Level.java b/src/main/java/net/minecraft/world/level/Level.java
index b2e23a398fa53de1e40b406379a959244d4200dc..98b69acf7691c79bfae86d7415b30916c9370261 100644
--- a/src/main/java/net/minecraft/world/level/Level.java
+++ b/src/main/java/net/minecraft/world/level/Level.java
@@ -254,6 +254,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable {
         if (ret == null) {
             return CachedWorldData.get(world.getName()); // DirtyFolia - cache world data
         }
+        CachedWorldData.update(world.getName(), ret);
         Level world = ret.world;
         if (world != this) {
             throw new IllegalStateException("World mismatch: expected " + this.getWorld().getName() + " but got " + world.getWorld().getName());
