From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: killerprojecte <admin@fastmcmirror.org>
Date: Mon, 3 Jul 2023 20:05:03 +0800
Subject: [PATCH] Fix ServerConnectionListener return empty connections


diff --git a/src/main/java/io/papermc/paper/threadedregions/RegionizedServer.java b/src/main/java/io/papermc/paper/threadedregions/RegionizedServer.java
index 7d620ac3b071415d55518fcea3fa634eb8e07967..d76e391f80d3aa2227c39d099316936e4933c6fe 100644
--- a/src/main/java/io/papermc/paper/threadedregions/RegionizedServer.java
+++ b/src/main/java/io/papermc/paper/threadedregions/RegionizedServer.java
@@ -50,6 +50,10 @@ public final class RegionizedServer {
         this.connections.add(conn);
     }
 
+    public List<Connection> getConnections() {
+        return this.connections; // DirtyFolia - Fix ServerConnectionListener get empty connection list
+    }
+
     private boolean removeConnection(final Connection conn) {
         return this.connections.remove(conn);
     }
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 89d4ae824d7786402819abbd9084a0d9320a4023..6b259a38522c4d09319eba3d73a4863e8b69794f 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1500,6 +1500,8 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             }
             // now tick connections
             region.world.getCurrentWorldData().tickConnections(); // Folia - region threading
+            // DirtyFolia cache work data
+            CachedWorldData.update(region.world.serverLevelData.getLevelName(), region.world.getCurrentWorldData());
         }
         // Folia end - region threading
 
diff --git a/src/main/java/net/minecraft/server/network/ServerConnectionListener.java b/src/main/java/net/minecraft/server/network/ServerConnectionListener.java
index 072634e26d32ca0b3438a5d3a03be3670d0f846e..626b2765ca3fcd1718e97ce85a27f7b04abf232c 100644
--- a/src/main/java/net/minecraft/server/network/ServerConnectionListener.java
+++ b/src/main/java/net/minecraft/server/network/ServerConnectionListener.java
@@ -34,6 +34,8 @@ import java.util.Iterator;
 import java.util.List;
 import java.util.concurrent.TimeUnit;
 import javax.annotation.Nullable;
+
+import io.papermc.paper.threadedregions.RegionizedServer;
 import net.minecraft.CrashReport;
 import net.minecraft.ReportedException;
 import net.minecraft.network.Connection;
@@ -267,7 +269,7 @@ public class ServerConnectionListener {
     }
 
     public List<Connection> getConnections() {
-        return this.connections;
+        return RegionizedServer.getInstance().getConnections();
     }
 
     private static class LatencySimulator extends ChannelInboundHandlerAdapter {
